drivers/can/can_mcp251xfd.c

```diff
@@ -0,0 +1,1814 @@
+
+		data = mcp251xfd_read_reg(dev, memory_addr, len * fifo->item_size);
+		if (!data) {
+			LOG_ERR("Error fetching batch message");
+			ret = -EINVAL;
+			goto done;
+		}
+
+		for (int i = 0; i < len; i++) {
+			fifo->msg_handler(dev, (void *)(&data[i * fifo->item_size]));
+		}
+
+		fifo_tail_index = (fifo_tail_index + len) % fifo->capacity;
+		fetch_total -= len;
+		ui_inc += len;
+	}
+
+	reg_byte = mcp251xfd_get_spi_buf_ptr(dev);
+	*reg_byte = MCP251XFD_UINT32_FLAG_TO_BYTE_MASK(MCP251XFD_REG_FIFOCON_UINC);
+
+	for (int i = 0; i < ui_inc; i++) {
+		ret = mcp251xfd_write(dev, fifo->reg_fifocon_addr + 1, 1);
+		if (ret < 0) {
+			LOG_ERR("Failed to increment pointer");
+			goto done;
+		}
+	}
+
+done:
+	k_mutex_unlock(&dev_data->mutex);
+	return ret;
+}
+
+static void mcp251xfd_reset_tx_fifos(const struct device *dev, int status)
+{
+	struct mcp251xfd_data *dev_data = dev->data;
+
+	LOG_INF("All FIFOs Reset");
+	k_mutex_lock(&dev_data->mutex, K_FOREVER);
+	for (int i = 0; i < MCP251XFD_TX_QUEUE_ITEMS; i++) {
+		can_tx_callback_t callback;
+
+		if (!(dev_data->mailbox_usage & BIT(i))) {
+			continue;
+		}
+
+		callback = dev_data->mailbox[i].cb;
+		if (callback) {
+			callback(dev, status, dev_data->mailbox[i].cb_arg);
+		}
+
+		dev_data->mailbox_usage &= ~BIT(i);
+		dev_data->mailbox[i].cb = NULL;
+		k_sem_give(&dev_data->tx_sem);
+	}
+	k_mutex_unlock(&dev_data->mutex);
+}
+
+/*
+ * CERRIF will be set each time a threshold in the TEC/REC counter is crossed by the following
+ * conditions:
+ * • TEC or REC exceeds the Error Warning state threshold
+ * • The transmitter or receiver transitions to Error Passive state
+ * • The transmitter transitions to Bus Off state
+ * • The transmitter or receiver transitions from Error Passive to Error Active state
+ * • The module transitions from Bus Off to Error Active state, after the bus off recovery
+ * sequence
+ * When the user clears CERRIF, it will remain clear until a new counter crossing occurs.
+ */
+static int mcp251xfd_handle_cerrif(const struct device *dev)
+{
+	enum can_state new_state;
+	struct mcp251xfd_data *dev_data = dev->data;
+	uint32_t *reg;
+	uint8_t tec, rec;
+	int ret = 0;
+
+	k_mutex_lock(&dev_data->mutex, K_FOREVER);
+
+	reg = mcp251xfd_read_crc(dev, MCP251XFD_REG_TREC, MCP251XFD_REG_SIZE);
+	if (!reg) {
+		ret = -EINVAL;
+		goto done;
+	}
+
+	*reg = sys_le32_to_cpu(*reg);
+
+	if ((*reg & MCP251XFD_REG_TREC_TXBO) != 0) {
+		new_state = CAN_STATE_BUS_OFF;
+	} else if ((*reg & MCP251XFD_REG_TREC_TXBP) != 0) {
+		new_state = CAN_STATE_ERROR_PASSIVE;
+	} else if ((*reg & MCP251XFD_REG_TREC_RXBP) != 0) {
+		new_state = CAN_STATE_ERROR_PASSIVE;
+	} else if ((*reg & MCP251XFD_REG_TREC_RXWARN) != 0) {
+		new_state = CAN_STATE_ERROR_WARNING;
+	} else {
+		new_state = CAN_STATE_ERROR_ACTIVE;
+	}
+
+	if (new_state == dev_data->state) {
+		goto done;
+	}
+
+	tec = FIELD_GET(MCP251XFD_REG_TREC_TEC_MASK, *reg);
+	rec = FIELD_GET(MCP251XFD_REG_TREC_REC_MASK, *reg);
```

Comment:
Instead of parsing these two in tow different functions, could `mcp251xfd_get_state()` be called here instead?